protostar/git/git_repository.py

```diff
@@ -120,122 +80,47 @@ def add_submodule(
-            ],
-            **SHARED_KWARGS,
-            cwd=self.repo_path,
-        )
+    def remove_submodule(self, submodule_path: Path) -> None:
+        self._git(["rm", "--force", str(submodule_path)])
 
-    @wrap_git_exception
-    def commit(self, msg: str):
-        credentials = [] if has_git_credentials() else DEFAULT_CREDENTIALS
-        subprocess.run(
-            ["git", *credentials, "commit", "-m", msg],
-            **SHARED_KWARGS,
-            cwd=self.repo_path,
-        )
+    def commit(self, msg: str) -> None:
+        self._git(["commit", "-m", msg])
 
-    @wrap_git_exception
-    def checkout(self, branch: str):
-        subprocess.run(["git", "checkout", branch], **SHARED_KWARGS, cwd=self.repo_path)
+    def checkout(self, branch: str) -> None:
+        self._git(["checkout", branch])
 
-    @wrap_git_exception
-    def create_tag(self, name: str):
-        credentials = [] if has_git_credentials() else DEFAULT_CREDENTIALS
-        subprocess.run(
-            ["git", *credentials, "tag", name], **SHARED_KWARGS, cwd=self.repo_path
-        )
+    def create_tag(self, name: str) -> None:
+        self._git(["tag", name])
 
-    @wrap_git_exception
     def get_tag_rev(self) -> str:
-        process = subprocess.run(
-            ["git", "rev-list", "--tags", "--max-count=1"],
-            stderr=subprocess.DEVNULL,
-            stdout=subprocess.PIPE,
-            check=True,
-            cwd=self.repo_path,
-        )
-        return process.stdout.decode("utf-8").strip()
+        return self._git(["rev-list", "--tags", "--max-count=1"])
 
-    @wrap_git_exception
     def get_tag(self, rev: Optional[str] = None) -> str:
-        process = subprocess.run(
-            ["git", "describe", "--tags"] + ([rev] if rev else []),
-            stderr=subprocess.DEVNULL,
-            stdout=subprocess.PIPE,
-            check=True,
-            cwd=self.repo_path,
-        )
-        return process.stdout.decode("utf-8").strip()
+        return self._git(["describe", "--tags"] + ([rev] if rev else []))
 
-    @wrap_git_exception
     def get_head(self) -> str:
-        process = subprocess.run(
-            ["git", "rev-parse", "HEAD"],
-            stderr=subprocess.DEVNULL,
-            stdout=subprocess.PIPE,
-            check=True,
-            cwd=self.repo_path,
-        )
-        return process.stdout.decode("utf-8").strip()
+        return self._git(["rev-parse", "HEAD"])
 
-    @wrap_git_exception
     def fetch_tags(self):
-        credentials = [] if has_git_credentials() else DEFAULT_CREDENTIALS
-        subprocess.run(
-            ["git", *credentials, "fetch", "--tags"],
-            **SHARED_KWARGS,
-            cwd=self.repo_path,
-        )
+        self._git(["fetch", "--tags"])
 
-    def get_submodules(self) -> Dict[str, Submodule]:
-        """
-        Returns a dictionary of form:
-        submodule_name: PackageInfo
-        """
+    def _git(self, args: list[str]) -> str:
```

Comment:
[nitpick]: I think it is a good case for starred-expression arguments. Although this one works as well, so i'll leave it up to you
```suggestion
    def _git(self, *args: str) -> str:
```

Reply:
Also we could allow a broader spectrum of arguments like: `Union[str, int, Path]` and do the conversion inside the method

Reply:
[make git fn variadic](https://github.com/software-mansion/protostar/pull/1455/commits/f03b7783169afd37fe8840c81a22368fb7c3faf9)